cirq/value/tensor_product_state.py

```diff
@@ -0,0 +1,346 @@
+
+    def __getitem__(self, qubit):
+        """Return the NamedState at the given qubit."""
+        return self.states[qubit]
+
+    def __iter__(self):
+        yield from self.states.items()
+
+    def __len__(self):
+        return len(self.states)
+
+    def __eq__(self, other):
+        if not isinstance(other, TensorProductState):
+            return False
+
+        return self.states == other.states
+
+    def __hash__(self):
+        return hash(tuple(self.states.items()))
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': self.__class__.__name__,
+            'states': list(self.states.items()),
+        }
+
+    @classmethod
+    def _from_json_dict_(cls, states, **kwargs):
+        return cls(states=dict(states))
+
+    def state_vector(self, qubit_order: 'cirq.QubitOrder' = None):
+        """The state-vector representation of this state."""
+        from cirq import ops
+        if qubit_order is None:
+            qubit_order = ops.QubitOrder.DEFAULT
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(self.qubits)
+
+        mat = 1.0 + 0.j
+        for qubit in qubits:
+            oneq_state = self[qubit]
+            state_vector = oneq_state.state_vector()
+            mat = np.kron(mat, state_vector)
+
+        return mat
+
+    def projector(self, qubit_order: 'cirq.QubitOrder' = None):
+        """The projector associated with this state expressed as a matrix.
+
+        This is |s⟩⟨s| where |s⟩ is this state.
+        """
+        from cirq import ops
+        if qubit_order is None:
+            qubit_order = ops.QubitOrder.DEFAULT
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(self.qubits)
+
+        mat = 1.0 + 0.j
+        for qubit in qubits:
+            oneq_state = self[qubit]
+            oneq_proj = oneq_state.projector()
+            mat = np.kron(mat, oneq_proj)
+        return mat
+
+
+class _KetPlus(_NamedOneQubitState):
+
+    def __str__(self):
+        return '+X'
+
+    def __repr__(self):
+        return 'cirq.KET_PLUS'
+
+    def state_vector(self):
+        return np.array([1, 1]) / np.sqrt(2)
+
+    def stabilized_by(self):
+        # Prevent circular import from `value.value_equality`
+        from cirq import ops
+        return 1, ops.X
+
+
+class _KetMinus(_NamedOneQubitState):
+
+    def __str__(self):
+        return '-X'
+
+    def __repr__(self):
+        return 'cirq.KET_MINUS'
+
+    def state_vector(self):
+        return np.array([1, -1]) / np.sqrt(2)
+
+    def stabilized_by(self):
+        from cirq import ops
+        return -1, ops.X
+
+
+class _KetImag(_NamedOneQubitState):
+
+    def __str__(self):
+        return '+Y'
+
+    def __repr__(self):
+        return 'cirq.KET_IMAG'
+
+    def state_vector(self):
+        return np.array([1, 1j]) / np.sqrt(2)
+
+    def stabilized_by(self):
+        from cirq import ops
+        return 1, ops.Y
```

Comment:
The more I think about it the more I'm convinced that we should avoid defining a class to represent every interesting quantum state. Instead, we should define a class for each interesting basis.

The core of the argument is that all states in a given basis are so similar that the classes end up nearly identical. Repeating them is unnecessary and results in verbosity. Admittedly, verbosity inside the library is a minor issue. In this case, however, the building blocks we offer are so low level that the client code ends up verbose, too.

For example, suppose we wish to represent the state stabilized by XII, IYI and -IIZ. Using the code in this PR, we would say

```python
qs = cirq.LineQubit.range(3)
state = cirq.KET_PLUS(qs[0]) * cirq.KET_IMAG(qs[1]) * cirq.KET_ONE(qs[2])
```

A possible alternative may be:

```python
qs = cirq.LineQubit.range(3)
state = cirq.X.basis[+1] * cirq.Y.basis[+1] * cirq.Z.basis[-1]
```

where I assumed that:
1. We represent each basis as a class.
2. Each basis class is accessed via an operator using the ".basis" property.
3. The class supports indexing where the index is the corresponding eigenvalue.

The code is not just shorter, but is more amenable to programmatic generation of states, e.g.:

```python
def func(operators, eigenvalue):
    state = 1
    for op in operators:
        state *= op.basis[eigenvalue]
    return state

s = func([cirq.X, cirq.Y, cirq.Z], -1)
```

Also, the choice to use existing gates such as cirq.X provides a familiar hook for this functionality: once the user learns the ".basis" trick they know how to easily build complicated states without having to remember names we invent for the various states (which as the set of states grows are bound to become increasingly weird and idiosyncratic).

Side note: Interestingly, it seems that `EigenGate._eigen_components ` could be used readily to implement the `projector` member function that you need.

Reply:
what is the type of `x.basis[i]`? This looks nifty but looks like an alternative way of constructing OneQubitStates and does not obviate the need for such a class

Reply:
Ah, I was imprecise. I meant a situation where we'd have one class per interesting basis, but each instance of a basis class would represent a basis element (just like your instances do presently). In particular, it would be appropriate to name these classes as `*State` as you do currently (`__init__` arguments would distinguish basis elements). For example, in place of two classes such as `_KetPlus` and `_KetMinus`, we'd have a single class, say `_XEigenState` (and there would be two instances of this class). This reduces verbosity by a factor of two for qubits (and more for qudits).

Reply:
Yes, I think this is a good happy medium among all the various levels of verbosity that have been proposed here! I'll work on this change